Column-oriented data analysis API

In [2]:
import pandas as pd
pd.__version__

'1.5.3'

In [3]:
from sklearn.datasets import load_diabetes
diabetes = load_diabetes(as_frame = True)

print(type(diabetes['data']))

<class 'pandas.core.frame.DataFrame'>


In [4]:
df = diabetes['data']

Creating pandas dataframe from series

In [9]:
cities = pd.Series(['Mumbai', 'Banglore', 'Chennai', 'Delhi'])
population = pd.Series([17000000, 13000000, 6000000])

In [11]:
city_info_df = pd.DataFrame({'City': cities, 'Population': population})

In [12]:
print(city_info_df)
print(type(city_info_df))

       City  Population
0    Mumbai  17000000.0
1  Banglore  13000000.0
2   Chennai   6000000.0
3     Delhi         NaN
<class 'pandas.core.frame.DataFrame'>


We want to know what is there in these dataFrames if we are given one.

## Exploring Data in DataFrame

In [13]:
#Quickly examine a few entries in the dataframe. Say first 5 or last 5

df.head()

age       sex       bmi        bp        s1        s2        s3  \
0  0.038076  0.050680  0.061696  0.021872 -0.044223 -0.034821 -0.043401   
1 -0.001882 -0.044642 -0.051474 -0.026328 -0.008449 -0.019163  0.074412   
2  0.085299  0.050680  0.044451 -0.005670 -0.045599 -0.034194 -0.032356   
3 -0.089063 -0.044642 -0.011595 -0.036656  0.012191  0.024991 -0.036038   
4  0.005383 -0.044642 -0.036385  0.021872  0.003935  0.015596  0.008142   

         s4        s5        s6  
0 -0.002592  0.019907 -0.017646  
1 -0.039493 -0.068332 -0.092204  
2 -0.002592  0.002861 -0.025930  
3  0.034309  0.022688 -0.009362  
4 -0.002592 -0.031988 -0.046641

In [14]:
df.tail()

age       sex       bmi        bp        s1        s2        s3  \
437  0.041708  0.050680  0.019662  0.059744 -0.005697 -0.002566 -0.028674   
438 -0.005515  0.050680 -0.015906 -0.067642  0.049341  0.079165 -0.028674   
439  0.041708  0.050680 -0.015906  0.017293 -0.037344 -0.013840 -0.024993   
440 -0.045472 -0.044642  0.039062  0.001215  0.016318  0.015283 -0.028674   
441 -0.045472 -0.044642 -0.073030 -0.081413  0.083740  0.027809  0.173816   

           s4        s5        s6  
437 -0.002592  0.031193  0.007207  
438  0.034309 -0.018114  0.044485  
439 -0.011080 -0.046883  0.015491  
440  0.026560  0.044529 -0.025930  
441 -0.039493 -0.004222  0.003064

in case we do not know anything about a function, we can use `?` before the function

In [15]:
? df.head #DOCUMENTATION WILL OPEN UP

In [17]:
#YOU CAN ALSO MENTION NUMBER OF ROWS YOU WANT TO DISPLAY

df.head(n=13)

age       sex       bmi        bp        s1        s2        s3  \
0   0.038076  0.050680  0.061696  0.021872 -0.044223 -0.034821 -0.043401   
1  -0.001882 -0.044642 -0.051474 -0.026328 -0.008449 -0.019163  0.074412   
2   0.085299  0.050680  0.044451 -0.005670 -0.045599 -0.034194 -0.032356   
3  -0.089063 -0.044642 -0.011595 -0.036656  0.012191  0.024991 -0.036038   
4   0.005383 -0.044642 -0.036385  0.021872  0.003935  0.015596  0.008142   
5  -0.092695 -0.044642 -0.040696 -0.019442 -0.068991 -0.079288  0.041277   
6  -0.045472  0.050680 -0.047163 -0.015999 -0.040096 -0.024800  0.000779   
7   0.063504  0.050680 -0.001895  0.066629  0.090620  0.108914  0.022869   
8   0.041708  0.050680  0.061696 -0.040099 -0.013953  0.006202 -0.028674   
9  -0.070900 -0.044642  0.039062 -0.033213 -0.012577 -0.034508 -0.024993   
10 -0.096328 -0.044642 -0.083808  0.008101 -0.103389 -0.090561 -0.013948   
11  0.027178  0.050680  0.017506 -0.033213 -0.007073  0.045972 -0.065491   
12  0.016281 -0.044642 -0.028840 -0.009113 -0.004321 -0.009769  0.044958   

          s4        s5        s6  
0  -0.002592  0.019907 -0.017646  
1  -0.039493 -0.068332 -0.092204  
2  -0.002592  0.002861 -0.025930  
3   0.034309  0.022688 -0.009362  
4  -0.002592 -0.031988 -0.046641  
5  -0.076395 -0.041176 -0.096346  
6  -0.039493 -0.062917 -0.038357  
7   0.017703 -0.035816  0.003064  
8  -0.002592 -0.014960  0.011349  
9  -0.002592  0.067737 -0.013504  
10 -0.076395 -0.062917 -0.034215  
11  0.071210 -0.096435 -0.059067  
12 -0.039493 -0.030748 -0.042499

In [19]:
#USE ATTRIBUTE OF THE DATAFRAME 'shape' to know shape of the dataframe

df.shape

(442, 10)

in this case there are 442 rows and 10 columns

We may want to know the names of the columns

In [21]:
df.columns

Index(['age', 'sex', 'bmi', 'bp', 's1', 's2', 's3', 's4', 's5', 's6'], dtype='object')

convert it into list format

In [22]:
list(df.columns)

['age', 'sex', 'bmi', 'bp', 's1', 's2', 's3', 's4', 's5', 's6']

In [24]:
print(diabetes.DESCR)

.. _diabetes_dataset:

Diabetes dataset
----------------

Ten baseline variables, age, sex, body mass index, average blood
pressure, and six blood serum measurements were obtained for each of n =
442 diabetes patients, as well as the response of interest, a
quantitative measure of disease progression one year after baseline.

**Data Set Characteristics:**

  :Number of Instances: 442

  :Number of Attributes: First 10 columns are numeric predictive values

  :Target: Column 11 is a quantitative measure of disease progression one year after baseline

  :Attribute Information:
      - age     age in years
      - sex
      - bmi     body mass index
      - bp      average blood pressure
      - s1      tc, total serum cholesterol
      - s2      ldl, low-density lipoproteins
      - s3      hdl, high-density lipoproteins
      - s4      tch, total cholesterol / HDL
      - s5      ltg, possibly log of serum triglycerides level
      - s6      glu, blood sugar level

Note: Each of these 1

In [25]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 442 entries, 0 to 441
Data columns (total 10 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   age     442 non-null    float64
 1   sex     442 non-null    float64
 2   bmi     442 non-null    float64
 3   bp      442 non-null    float64
 4   s1      442 non-null    float64
 5   s2      442 non-null    float64
 6   s3      442 non-null    float64
 7   s4      442 non-null    float64
 8   s5      442 non-null    float64
 9   s6      442 non-null    float64
dtypes: float64(10)
memory usage: 34.7 KB


All float columns, all non-null columns, and memory usage of the dataframe is 34.7KB


In [26]:
df.describe()

age           sex           bmi            bp            s1  \
count  4.420000e+02  4.420000e+02  4.420000e+02  4.420000e+02  4.420000e+02   
mean  -2.511817e-19  1.230790e-17 -2.245564e-16 -4.797570e-17 -1.381499e-17   
std    4.761905e-02  4.761905e-02  4.761905e-02  4.761905e-02  4.761905e-02   
min   -1.072256e-01 -4.464164e-02 -9.027530e-02 -1.123988e-01 -1.267807e-01   
25%   -3.729927e-02 -4.464164e-02 -3.422907e-02 -3.665608e-02 -3.424784e-02   
50%    5.383060e-03 -4.464164e-02 -7.283766e-03 -5.670422e-03 -4.320866e-03   
75%    3.807591e-02  5.068012e-02  3.124802e-02  3.564379e-02  2.835801e-02   
max    1.107267e-01  5.068012e-02  1.705552e-01  1.320436e-01  1.539137e-01   

                 s2            s3            s4            s5            s6  
count  4.420000e+02  4.420000e+02  4.420000e+02  4.420000e+02  4.420000e+02  
mean   3.918434e-17 -5.777179e-18 -9.042540e-18  9.293722e-17  1.130318e-17  
std    4.761905e-02  4.761905e-02  4.761905e-02  4.761905e-02  4.761905e-02  
min   -1.156131e-01 -1.023071e-01 -7.639450e-02 -1.260971e-01 -1.377672e-01  
25%   -3.035840e-02 -3.511716e-02 -3.949338e-02 -3.324559e-02 -3.317903e-02  
50%   -3.819065e-03 -6.584468e-03 -2.592262e-03 -1.947171e-03 -1.077698e-03  
75%    2.984439e-02  2.931150e-02  3.430886e-02  3.243232e-02  2.791705e-02  
max    1.987880e-01  1.811791e-01  1.852344e-01  1.335973e-01  1.356118e-01

In [27]:
df.describe(percentiles = [0.2, 0.6, 0.8])

age           sex           bmi            bp            s1  \
count  4.420000e+02  4.420000e+02  4.420000e+02  4.420000e+02  4.420000e+02   
mean  -2.511817e-19  1.230790e-17 -2.245564e-16 -4.797570e-17 -1.381499e-17   
std    4.761905e-02  4.761905e-02  4.761905e-02  4.761905e-02  4.761905e-02   
min   -1.072256e-01 -4.464164e-02 -9.027530e-02 -1.123988e-01 -1.267807e-01   
20%   -4.547248e-02 -4.464164e-02 -4.048038e-02 -4.009893e-02 -3.871969e-02   
50%    5.383060e-03 -4.464164e-02 -7.283766e-03 -5.670422e-03 -4.320866e-03   
60%    1.628068e-02  5.068012e-02  5.218854e-03  8.100982e-03  8.062710e-03   
80%    4.170844e-02  5.068012e-02  4.229559e-02  4.941519e-02  3.943444e-02   
max    1.107267e-01  5.068012e-02  1.705552e-01  1.320436e-01  1.539137e-01   

                 s2            s3            s4            s5            s6  
count  4.420000e+02  4.420000e+02  4.420000e+02  4.420000e+02  4.420000e+02  
mean   3.918434e-17 -5.777179e-18 -9.042540e-18  9.293722e-17  1.130318e-17  
std    4.761905e-02  4.761905e-02  4.761905e-02  4.761905e-02  4.761905e-02  
min   -1.156131e-01 -1.023071e-01 -7.639450e-02 -1.260971e-01 -1.377672e-01  
20%   -3.695017e-02 -3.971921e-02 -3.949338e-02 -4.117617e-02 -3.835666e-02  
50%   -3.819065e-03 -6.584468e-03 -2.592262e-03 -1.947171e-03 -1.077698e-03  
60%    8.706873e-03  8.142084e-03 -2.592262e-03  1.255119e-02  7.206516e-03  
80%    3.952068e-02  3.759519e-02  3.430886e-02  3.885335e-02  3.620126e-02  
max    1.987880e-01  1.811791e-01  1.852344e-01  1.335973e-01  1.356118e-01

you can also transpose this

In [28]:
df.describe().T

count          mean       std       min       25%       50%       75%  \
age  442.0 -2.511817e-19  0.047619 -0.107226 -0.037299  0.005383  0.038076   
sex  442.0  1.230790e-17  0.047619 -0.044642 -0.044642 -0.044642  0.050680   
bmi  442.0 -2.245564e-16  0.047619 -0.090275 -0.034229 -0.007284  0.031248   
bp   442.0 -4.797570e-17  0.047619 -0.112399 -0.036656 -0.005670  0.035644   
s1   442.0 -1.381499e-17  0.047619 -0.126781 -0.034248 -0.004321  0.028358   
s2   442.0  3.918434e-17  0.047619 -0.115613 -0.030358 -0.003819  0.029844   
s3   442.0 -5.777179e-18  0.047619 -0.102307 -0.035117 -0.006584  0.029312   
s4   442.0 -9.042540e-18  0.047619 -0.076395 -0.039493 -0.002592  0.034309   
s5   442.0  9.293722e-17  0.047619 -0.126097 -0.033246 -0.001947  0.032432   
s6   442.0  1.130318e-17  0.047619 -0.137767 -0.033179 -0.001078  0.027917   

          max  
age  0.110727  
sex  0.050680  
bmi  0.170555  
bp   0.132044  
s1   0.153914  
s2   0.198788  
s3   0.181179  
s4   0.185234  
s5   0.133597  
s6   0.135612

##Selection

In [29]:
df.columns

Index(['age', 'sex', 'bmi', 'bp', 's1', 's2', 's3', 's4', 's5', 's6'], dtype='object')

In [30]:
df['age']

0      0.038076
1     -0.001882
2      0.085299
3     -0.089063
4      0.005383
         ...   
437    0.041708
438   -0.005515
439    0.041708
440   -0.045472
441   -0.045472
Name: age, Length: 442, dtype: float64

In [31]:
df['age'][2]

0.08529890629667548

In [32]:
type(df['age'])

pandas.core.series.Series

In [33]:
df['age'][0]

0.038075906433423026

Select first 5 values

In [34]:
df['age'][:5]

0    0.038076
1   -0.001882
2    0.085299
3   -0.089063
4    0.005383
Name: age, dtype: float64

select last 5 values

In [35]:
df['age'][-5:]

437    0.041708
438   -0.005515
439    0.041708
440   -0.045472
441   -0.045472
Name: age, dtype: float64

interval selection

In [36]:
df['age'][100:200]

100    0.016281
101    0.016281
102   -0.092695
103    0.059871
104   -0.027310
         ...   
195    0.027178
196   -0.023677
197    0.048974
198   -0.052738
199    0.041708
Name: age, Length: 100, dtype: float64

for selecting two more columns, we have to give a list of columns

In [37]:
df[['age','sex']]

#Selecting first 5 columns
df[['age','sex']][:5]

#Selecting last 5 columns
df[['age','sex']][-5:]


age       sex
0    0.038076  0.050680
1   -0.001882 -0.044642
2    0.085299  0.050680
3   -0.089063 -0.044642
4    0.005383 -0.044642
..        ...       ...
437  0.041708  0.050680
438 -0.005515  0.050680
439  0.041708  0.050680
440 -0.045472 -0.044642
441 -0.045472 -0.044642

[442 rows x 2 columns]

Let's discuss now, `.loc` and `iloc`

In [38]:
#df[0] will not work

df.iloc[0] #PRINTS THE FIRST ROW

age    0.038076
sex    0.050680
bmi    0.061696
bp     0.021872
s1    -0.044223
s2    -0.034821
s3    -0.043401
s4    -0.002592
s5     0.019907
s6    -0.017646
Name: 0, dtype: float64

In [41]:
df.iloc[441] #PRINTS THE 441st ROW

age   -0.045472
sex   -0.044642
bmi   -0.073030
bp    -0.081413
s1     0.083740
s2     0.027809
s3     0.173816
s4    -0.039493
s5    -0.004222
s6     0.003064
Name: 441, dtype: float64

`iloc` specifies the index of the row (only accepts integers values)

whereas

`loc` specifies the label of the row (also accepts strings along with integer values)

In [43]:
df.loc[0]

age    0.038076
sex    0.050680
bmi    0.061696
bp     0.021872
s1    -0.044223
s2    -0.034821
s3    -0.043401
s4    -0.002592
s5     0.019907
s6    -0.017646
Name: 0, dtype: float64

In [44]:
df.loc[4, 'age']

0.005383060374248237

In [45]:
df.iloc[4, 0] # GET ME 4th VALUE IN FIRST COLUMN

0.005383060374248237

In [46]:
df.iloc[4, 0:2] #GET ME 4th RECORD OF FIRST TWO COLUMNS

age    0.005383
sex   -0.044642
Name: 4, dtype: float64

In [47]:
df.loc[4, ['age', 'sex']]

age    0.005383
sex   -0.044642
Name: 4, dtype: float64

In [48]:
df.iloc[4, [1,5,7,9]]

sex   -0.044642
s2     0.015596
s4    -0.002592
s6    -0.046641
Name: 4, dtype: float64

condition

In [50]:
age_df_temp = df.loc[df.age < 5.38]
age_df_temp

age       sex       bmi        bp        s1        s2        s3  \
0    0.038076  0.050680  0.061696  0.021872 -0.044223 -0.034821 -0.043401   
1   -0.001882 -0.044642 -0.051474 -0.026328 -0.008449 -0.019163  0.074412   
2    0.085299  0.050680  0.044451 -0.005670 -0.045599 -0.034194 -0.032356   
3   -0.089063 -0.044642 -0.011595 -0.036656  0.012191  0.024991 -0.036038   
4    0.005383 -0.044642 -0.036385  0.021872  0.003935  0.015596  0.008142   
..        ...       ...       ...       ...       ...       ...       ...   
437  0.041708  0.050680  0.019662  0.059744 -0.005697 -0.002566 -0.028674   
438 -0.005515  0.050680 -0.015906 -0.067642  0.049341  0.079165 -0.028674   
439  0.041708  0.050680 -0.015906  0.017293 -0.037344 -0.013840 -0.024993   
440 -0.045472 -0.044642  0.039062  0.001215  0.016318  0.015283 -0.028674   
441 -0.045472 -0.044642 -0.073030 -0.081413  0.083740  0.027809  0.173816   

           s4        s5        s6  
0   -0.002592  0.019907 -0.017646  
1   -0.039493 -0.068332 -0.092204  
2   -0.002592  0.002861 -0.025930  
3    0.034309  0.022688 -0.009362  
4   -0.002592 -0.031988 -0.046641  
..        ...       ...       ...  
437 -0.002592  0.031193  0.007207  
438  0.034309 -0.018114  0.044485  
439 -0.011080 -0.046883  0.015491  
440  0.026560  0.044529 -0.025930  
441 -0.039493 -0.004222  0.003064  

[442 rows x 10 columns]

In [55]:
age_df_temp = df.loc[(df['age'] < 5.38) & (df['sex'] == 1)]
